tardis/energy_input/base.py

```diff
@@ -0,0 +1,369 @@
+                    )
+
+                    # annihilation produces second gamma-ray in opposite direction
+                    (
+                        ray1_x,
+                        ray1_y,
+                        ray1_z,
+                    ) = ray1.direction.cartesian_coords
+                    ray2 = copy.deepcopy(ray1)
+                    ray2_x = -ray1_x
+                    ray2_y = -ray1_y
+                    ray2_z = -ray1_z
+                    ray2_r, ray2_theta, ray2_phi = cartesian_to_spherical(
+                        ray2_x, ray2_y, ray2_z
+                    )
+                    ray2.direction.r = ray2_r
+                    ray2.direction.theta = ray2_theta.value + 0.5 * np.pi
+                    ray2.direction.phi = ray2_phi.value
+
+                else:
+                    # Spawn a gamma ray emission with energy from gamma-ray list
+                    ray1.energy = sample_energy_distribution(
+                        energy_sorted, energy_cdf
+                    )
+                    packets.append(ray1)
+
+    return packets, energy_df_rows, energy_plot_df_rows
+
+
+def main_gamma_ray_loop(num_packets, model):
+    """Main loop that determines the gamma ray propagation
+
+    Parameters
+    ----------
+    num_packets : int
+        Number of packets
+    model : tardis.Radial1DModel
+        The tardis model to calculate gamma ray propagation through
+
+    Returns
+    -------
+    pandas DataFrame
+        Energy per mass per shell in units of erg / g
+    pandas DataFrame
+        Columns:
+        Packet index,
+        Energy input per packet,
+        radius of deposition,
+        theta angle of deposition,
+        time of deposition,
+        type of deposition where:
+            -1 = beta decay,
+            0 = Compton scatter,
+            1 = photoabsorption,
+            2 = pair creation
+    list
+        Energy of escaping packets
+    """
+    escape_energy = []
+    interaction_count = []
+
+    # Note the use of velocity as the radial coordinate
+    inner_radius = model.v_inner[0].value
+    outer_radius = model.v_outer[-1].value
+    outer_velocities = model.v_outer[:].value
+    inner_velocities = model.v_inner[:].value
+    ejecta_density = model.density[:].value
+    ejecta_volume = model.volume[:].value
+    ejecta_epoch = model.time_explosion.to("s").value
+    number_of_shells = model.no_of_shells
+    raw_isotope_abundance = model.raw_isotope_abundance
+
+    shell_masses = ejecta_volume * ejecta_density
+
+    (
+        packets_per_shell,
+        decay_rad_db,
+        decay_rate_per_shell,
+    ) = compute_required_packets_per_shell(
+        shell_masses,
+        raw_isotope_abundance,
+        num_packets,
+    )
+
+    scaled_decay_rate_per_shell = (
+        decay_rate_per_shell / packets_per_shell.to_numpy().sum(axis=1)
+    )
+
+    # Taking iron group to be elements 21-30
+    iron_group_fraction_per_shell = raw_isotope_abundance.loc[(21,):(30,)].sum(
```

Comment:
we should make sure that we document that this is a choice because of atomic data.